syntax = "proto3";

option go_package = "/pkg/genproto/backtests";

package backtests;

service BacktestsService {
    rpc CreateBacktest(CreateBacktestRequest) returns (CreateBacktestResponse) {}
    rpc SubscribeToBacktestEvents(SubscribeToBacktestEventsRequest) returns (SubscribeToBacktestEventsResponse){}
    rpc ListenBacktest(ListenBacktestRequest) returns (stream Event) {}
    rpc AdvanceBacktest(AdvanceBacktestRequest) returns (AdvanceBacktestResponse) {}
}

message AssetQuantity {
    string asset_name = 1;
    float quantity = 2;
}

message Account {
    string exchange_name = 1;
    repeated AssetQuantity assets = 2;
}

message CreateBacktestRequest {
    repeated Account accounts = 1;
	string start_time = 2;
	string end_time = 3;
    uint64 seconds_between_price_events = 4;
}

message CreateBacktestResponse {
    uint64 id = 1;
}

message SubscribeToBacktestEventsRequest {
    uint64 id = 1;
    string exchange = 2;
    string pair_symbol = 3;
}

message SubscribeToBacktestEventsResponse{
}

message AdvanceBacktestRequest {
    uint64 id = 1;
}

message AdvanceBacktestResponse {
    bool finished = 1;
}

message ListenBacktestRequest {
    uint64 id = 1;
}

message Event {
    string type = 1;
    string time = 2;
    string content = 3;
}
